### PR TITLE
Utility: Implement hasDarkSystray_private on Windows

### PR DESCRIPTION
**Pull request by TheOneRing**

### CODE REVIEW

**Comment by TheOneRing** [comment]
Naming and QWidget dep handling might be improved....

**Comment by TheOneRing** [comment]
Naming should be discussed 😅

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -2,7 +2,7 @@
 if(NOT TOKEN_AUTH_ONLY)
 endif()
 
-find_package(Qt5 5.6 COMPONENTS Core Network Xml Concurrent REQUIRED)
+find_package(Qt5 5.6 COMPONENTS Core Network Xml Concurrent Gui REQUIRED)
```
Comment (guruz) [inline src/CMakeLists.txt]: Hm I'm not sure about this one :-/

File: src/common/winrthelper/winrthelper.cpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+#include <Windows.UI.ViewManagement.h>
+#include <QColor>
+
+namespace  {
+
+QColor GetThemeColor(ABI::Windows::UI::ViewManagement::UIColorType type)
+{
+    // Error checking has been elided for expository purposes.
+    Microsoft::WRL::ComPtr<ABI::Windows::UI::ViewManagement::IUISettings3> settings;
+    if (SUCCEEDED(Windows::Foundation::ActivateInstance(Microsoft::WRL::Wrappers::HStringReference(
```
Comment (guruz) [inline src/common/winrthelper/winrthelper.cpp]: Is this a problem when executed on older windows versions I wonder?

Comment (TheOneRing) [reply src/common/winrthelper/winrthelper.cpp]: The code is com based so it should not make a problem, the only thing is the new library I link

File: src/CMakeLists.txt (hunk 1 above)
Comment (TheOneRing) [reply src/CMakeLists.txt]: Me neither, I just use it for QColor, could implement it manually but...
In general I'm not 100% happy how the utility code is handled here.

Comment (ckamm) [reply src/CMakeLists.txt]: We don't want libsync to depend on QtGui. I think this was a problem before. Possibly `Theme` should be split into a sync and a gui part, ideally in a way that makes ownbrander continue to work. 

You could look at doing that, but I guess that avoiding QColor here will be easier.

Comment (TheOneRing) [reply src/CMakeLists.txt]: I removed the dep to qtgui, and computate the lightness myself

**Comment by TheOneRing** [comment]
> 
> 
> Why do we need a new library, cant we just add winrthelper.cpp to csync?
> 
> I can't tell if linking against runtimeobject is allowed on all supported windows. I trust you on that.

I tried to make it even simpler by just setting the c++ standard for the utility_win.cpp to c++17 but it seems that has only an effect on targets, not source files.

**Comment by TheOneRing** [comment]
For the support of older versions of Windows, I'll try it in a vm. If it fails I could make the helper lib dynamic and try to load it during runtime?

**Comment by dschmidt** [comment]
> I tried to make it even simpler by just setting the c++ standard for the utility_win.cpp to c++17 but it seems that has only an effect on targets, not source files.

Out of curiosity what in your code does need C++17?

**Comment by TheOneRing** [comment]
> 
> 
> > I tried to make it even simpler by just setting the c++ standard for the utility_win.cpp to c++17 but it seems that has only an effect on targets, not source files.
> 
> Just out of curiosity: why do you need C++17?

Because the window rt headers require it....

**Comment by TheOneRing** [comment]
image
So a plugin or dynamic loaded lib?

**Comment by dschmidt** [comment]
> So a plugin or dynamic loaded lib?

Isn't a plugin a dynamically loaded lib? :-P 
If you mean QPlugin vs. a simple lib exporting a single function ... I don't mind.

**Comment by TheOneRing** [comment]
> 
> 
> > So a plugin or dynamic loaded lib?
> 
> Isn't a plugin a dynamically loaded lib? :-P
> If you mean QPlugin vs. a simple lib exporting a single function ... I don't mind.

Thats what I meant

**Comment by ogoffart** [comment]
> I tried to make it even simpler by just setting the c++ standard for the utility_win.cpp to c++17 but it seems that has only an effect on targets, not source files.

Would it not be even simpler to set the C++17 support to the whole project on windows?
If windows rt requires it anyway....

**Comment by TheOneRing** [comment]
Well in my experience c++14 is already quite progressive and I'm against different c++ standards for different platforms.
But bumping the requirements on linux is a bit harder, with all the distributions out there.

**Comment by TheOneRing** [comment]
My test was a custom build and is not representative.
I'll test it again with an official ci build, as far as I can see we would normally ship those libs.

**Comment by TheOneRing** [comment]
Tested on Windows 7 and 8.
Depends on Windows 8.

**Comment by guruz** [comment]
We could make it depend on the 64bit build and say that one will only have it.

**Comment by TheOneRing** [comment]
I would rather introduce an explicit option to set the minimum Windows version required

**Comment by guruz** [comment]
> I would rather introduce an explicit option to set the minimum Windows version required

What does that mean? Break old Windows versions?

**Comment by TheOneRing** [comment]
option(BUILD_WINRT_FEATURES...

**Comment by TheOneRing** [comment]
I fixed the issues by dynamically loading the component

File: changelog/unreleased/7356 [hunk 3]
```diff
@@ -0,0 +1,5 @@
+Change: Detect the Windows 10 thme for the system tray
```
Comment (dschmidt) [inline changelog/unreleased/7356:1]: *theme

File: src/common/utility_win.cpp [hunk 4]
```diff
@@ -29,10 +29,29 @@
...
 
-static const char systemRunPathC[] = "HKEY_LOCAL_MACHINE\\Software\\Microsoft\\Windows\\CurrentVersion\\Run";
-static const char runPathC[] = "HKEY_CURRENT_USER\\Software\\Microsoft\\Windows\\CurrentVersion\\Run";
+namespace {
+const char systemRunPathC[] = "HKEY_LOCAL_MACHINE\\Software\\Microsoft\\Windows\\CurrentVersion\\Run";
+const char runPathC[] = "HKEY_CURRENT_USER\\Software\\Microsoft\\Windows\\CurrentVersion\\Run";
+const QString rtHelperNameC()
+{
+    return QStringLiteral(APPLICATION_EXECUTABLE "_winrt");
+}
```
Comment (dschmidt) [inline src/common/utility_win.cpp:43]: could be a const char[] just like the defines above, no?

**Review by dschmidt — APPROVED** [review]

File: src/common/utility_win.cpp (hunk 4 above)
Comment (TheOneRing) [reply src/common/utility_win.cpp:43]: the first parameter from QLibaray takes a QString not char array, anyhow I will pass a absolute string to ensure we are loading the right lib...

**Pull request merged**